Fix mergeCollection bug

tgolen opened the pull request:
### Details
This fixes a bug found in mergeCollection

### Related Issues
https://github.com/Expensify/Expensify/issues/161384

### Automated Tests
There is a test added which properly tests the behavior

### Linked PRs
https://github.com/Expensify/Expensify.cash/pull/2543

tgolen commented:
I have read the CLA Document and I hereby sign the CLA

marcaaron commented on lib/Onyx.js:
```diff
@@ -195,7 +195,10 @@ function keysChanged(collectionKey, collection) {
+                    // If `newData` happens to not exist, then set the state back to the previous value because
+                    // if the state is set to an undefined value, then the item reverts to whatever is in
+                    // defaultProps
+                    [subscriber.statePropertyName]: typeof newData !== 'undefined' ? newData : prevState[subscriber.statePropertyName],
```
Should we use `_.isUndefined()` ?

tgolen replied:
Ah, yeah... that's a good call. I guess we do have underscore here.

marcaaron reviewed (commented):
Looks great and nice catch!

tgolen commented:
Updated to use underscore

marcaaron commented on lib/Onyx.js:
```diff
@@ -195,7 +195,12 @@ function keysChanged(collectionKey, collection) {
+                    // defaultProps
+                    [subscriber.statePropertyName]: !_.isUndefined(newData)
+                        ? newData
+                        : prevState[subscriber.statePropertyName],
```
Thinking more on this... this could maybe be filed under "unrelated" or "not a problem", but I find it interesting (or rather incorrect/inefficient) that if a subscriber is connected to a single key, but it's data hasn't changed at all ... that we are still calling `setState()` with previous data?

That's not super great because it means that we will still trigger re-renders on those connected components and for no reason at all. So, probably if we have `undefined` then we should do less than set the previous state again and should do... well... nothing?

tgolen replied:
Yeah, I think you make a good point. While it's not a problem currently, it's not a great design either. The reason I put it inside `setState` is because that's where it is getting `newData`. But now that I look at it closely, `newData` should be able to be retrieved before calling `setState()` and in the case where it is undefined, we can just skip calling `setState()`. How's that sound?

marcaaron replied:
Yep, sounds good to me. 👍

marcaaron commented on tests/unit/withOnyxTest.js:
```diff
@@ -153,3 +153,70 @@ describe('withOnyx', () => {
+                expect(onRender2.mock.calls[0][0].testObject).toStrictEqual({ID: 2});
+                expect(onRender2.mock.calls[1][0].testObject).toStrictEqual({ID: 2});
+                expect(onRender3.mock.calls[0][0].testObject).toStrictEqual({ID: 3});
+                expect(onRender3.mock.calls[1][0].testObject).toStrictEqual({ID: 3});
```
Like my expectation would be... not that this gets called again with same data but not called at all ? Why should it re-render since nothing has changed?

tgolen commented:
OK, updated. Now with less re-renderings!

chiragsalian reviewed (dismissed):
LGTM and i like the changes 👍

tgolen commented:
OK, before merging this, I want to give this a little more testing because I haven't been able to test the most recent changes yet. Once I do that testing, I'll let you know when this has a green-light to merge.

deetergp commented on tests/unit/withOnyxTest.js:
```diff
@@ -153,3 +153,73 @@ describe('withOnyx', () => {
+        const onRender2 = jest.fn();
+        const onRender3 = jest.fn();
+
+        // GIVEN there is a collection with three simple items in it
```
NAB: I REALLY appreciate the `GIVEN/WHEN/THEN` docs here! It makes it so much easier to follow what's going on and what we're testing for. `:bigfan:`

tgolen replied:
Yeah, I love that format for testing comments. @nkuoch was the one that originally showed it to me!

deetergp reviewed (dismissed):
Looks good to me

tgolen commented:
OK, I think I found another bug in my testing, but I want to write a test case to confirm that theory. Will continue this on Monday.

tgolen commented:
All right, I am taking off the hold here. When I tested this on Friday I found another bug with how the new collection data was being merged into the state. I have added a test case to cover that, fixed the bug (using spread operators instead of `lodash.merge()`), and I've tested this in the IOU PR to ensure it works the way it's supposed to.

chiragsalian reviewed (approved):
LGTM

marcaaron commented on lib/Onyx.js:
```diff
@@ -189,15 +189,22 @@ function keysChanged(collectionKey, collection) {
+            subscriber.withOnyxInstance.setState(prevState => ({
+                [subscriber.statePropertyName]: _.isObject(dataFromCollection)
+                    ? {
+                        ...prevState[subscriber.statePropertyName],
```
This seems potentially dangerous as it assumes we will always have at least an empty object for `prevState[subscriber.statePropertyName]` if we are dealing with object shaped data. If we try to spread something like `undefined` or `null` then we get 💥.

Maybe some combination of `lodash.merge()` could still be helpful here to help guard against that?

marcaaron replied:
Maybe there's no problem with that actually. I could just have made that up in my head.

marcaaron replied:
I think spreading `undefined` in objects works OK but arrays doesn't.

marcaaron replied:
I guess the small difference between this and using `lodashMerge()` is that the spread is not going to recursively merge the data. So depending on how complicated that data is we could get some weird results. But we can address that if it becomes a problem.

tgolen replied:
Cool, thanks Marc. I think it'll be OK for now. Seeing as how everyone has approved this, I'm going to go ahead and merge it.
